# Joining (Merging) DataFrames

Using the [MovieLens 100k data](http://grouplens.org/datasets/movielens/), let's create two DataFrames:

- **movies**: shows information about movies, namely a unique **movie_id** and its **title**
- **ratings**: shows the **rating** that a particular **user_id** gave to a particular **movie_id** at a particular **timestamp**

### Movies

In [1]:
import pandas as pd
movie_url = 'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.item'
movie_cols = ['movie_id', 'title']
movies = pd.read_table(movie_url, sep='|', header=None, names=movie_cols,
                       usecols=[0, 1], encoding="ISO-8859-1")
movies.head()


movie_id              title
0         1   Toy Story (1995)
1         2   GoldenEye (1995)
2         3  Four Rooms (1995)
3         4  Get Shorty (1995)
4         5     Copycat (1995)

In [2]:
#get the size of movies dataframe (number of rows )
len(movies)

1682

### Ratings

In [3]:
rating_url = 'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.data'
rating_cols = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_table(rating_url, sep='\t', header=None, names=rating_cols)
ratings.head()

user_id  movie_id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [4]:
len(ratings)

100000

Let's pretend that you want to examine the ratings DataFrame, but you want to know the **title** of each movie rather than its **movie_id**. The best way to accomplish this objective is by "joining" (or "merging") the DataFrames using the Pandas `merge` function:

In [5]:
movie_ratings = pd.merge(movies, ratings)
movie_ratings.head()
# movie_ratings = pd.merge(movies, ratings,
#                          left_on='movie_id',
#                          right_on='Movie_Id')


movie_id             title  user_id  rating  timestamp
0         1  Toy Story (1995)      308       4  887736532
1         1  Toy Story (1995)      287       5  875334088
2         1  Toy Story (1995)      148       4  877019411
3         1  Toy Story (1995)      280       4  891700426
4         1  Toy Story (1995)       66       3  883601324

In [6]:
pd.merge(ratings, movies)

user_id  movie_id  rating  timestamp  \
0          196       242       3  881250949   
1           63       242       3  875747190   
2          226       242       5  883888671   
3          154       242       3  879138235   
4          306       242       5  876503793   
5          296       242       4  884196057   
6           34       242       5  888601628   
7          271       242       4  885844495   
8          201       242       4  884110598   
9          209       242       4  883589606   
10          35       242       2  875459166   
11         354       242       5  891180399   
12         199       242       5  883782485   
13         113       242       2  875075887   
14           1       242       5  889751633   
15         173       242       5  877556626   
16         360       242       4  880353616   
17         234       242       4  891033261   
18          14       242       4  876964570   
19         309       242       4  877370319   
20         331       242       4  877196089   
21          21       242       3  874951617   
22         111       242       4  891679901   
23         439       242       5  882892424   
24         355       242       4  879486529   
25         204       242       5  892388935   
26         145       242       5  875269755   
27          30       242       5  885941156   
28         463       242       2  889935629   
29         144       242       4  888103444   
...        ...       ...     ...        ...   
99970      405      1564       1  885546288   
99971      456      1604       4  881372849   
99972      405      1574       1  885546529   
99973      883      1448       5  891695570   
99974      655      1448       3  887523224   
99975      234      1448       3  892335187   
99976      787      1671       1  888980193   
99977      587      1625       4  892871732   
99978      782      1665       2  891500194   
99979      181      1325       1  878962816   
99980      782      1670       3  891497793   
99981      181      1341       1  878962169   
99982      181      1352       1  878962240   
99983      655      1636       4  887473570   
99984      896      1681       3  887160722   
99985      655      1638       3  887488947   
99986      782      1667       3  891500110   
99987      851      1675       3  884222085   
99988      234      1460       3  892335460   
99989      532      1594       4  893115576   
99990      416      1594       5  893212484   
99991      648      1626       1  884795447   
99992      655      1645       4  892871225   
99993      747      1659       1  888733313   
99994      916      1682       3  880845755   
99995      840      1674       4  891211682   
99996      655      1640       3  888474646   
99997      655      1637       3  888984255   
99998      655      1630       3  887428735   
99999      655      1641       3  887427810   

                                                   title  
0                                           Kolya (1996)  
1                                           Kolya (1996)  
2                                           Kolya (1996)  
3                                           Kolya (1996)  
4                                           Kolya (1996)  
5                                           Kolya (1996)  
6                                           Kolya (1996)  
7                                           Kolya (1996)  
8                                           Kolya (1996)  
9                                           Kolya (1996)  
10                                          Kolya (1996)  
11                                          Kolya (1996)  
12                                          Kolya (1996)  
13                                          Kolya (1996)  
14                                          Kolya (1996)  
15                                          Kolya (1996)  
16                                          Kolya (1996)  
17                            

In [7]:
len(movie_ratings)


100000

In [8]:
ratings1 = movie_ratings.copy(deep=True)
ratings1 = ratings1[1:]
ratings1[0:10]


movie_id             title  user_id  rating  timestamp
1          1  Toy Story (1995)      287       5  875334088
2          1  Toy Story (1995)      148       4  877019411
3          1  Toy Story (1995)      280       4  891700426
4          1  Toy Story (1995)       66       3  883601324
5          1  Toy Story (1995)        5       4  875635748
6          1  Toy Story (1995)      109       4  880563619
7          1  Toy Story (1995)      181       3  878962392
8          1  Toy Story (1995)       95       5  879197329
9          1  Toy Story (1995)      268       3  875742341
10         1  Toy Story (1995)      189       5  893264174

In [9]:
ratings2 = movie_ratings.copy(deep=True)
ratings2[0:10]

#ratings[:-1]



movie_id             title  user_id  rating  timestamp
0         1  Toy Story (1995)      308       4  887736532
1         1  Toy Story (1995)      287       5  875334088
2         1  Toy Story (1995)      148       4  877019411
3         1  Toy Story (1995)      280       4  891700426
4         1  Toy Story (1995)       66       3  883601324
5         1  Toy Story (1995)        5       4  875635748
6         1  Toy Story (1995)      109       4  880563619
7         1  Toy Story (1995)      181       3  878962392
8         1  Toy Story (1995)       95       5  879197329
9         1  Toy Story (1995)      268       3  875742341

In [10]:
diff = ratings2['rating'] - ratings1['rating']

print(diff.head())

0    NaN
1    0.0
2    0.0
3    0.0
4    0.0
Name: rating, dtype: float64


Here's what just happened:

- Pandas noticed that movies and ratings had one column in common, namely **movie_id**. This is the "key" on which the DataFrames will be joined.
- The first **movie_id** in movies is 1. Thus, Pandas looked through every row in the ratings DataFrame, searching for a movie_id of 1. Every time it found such a row, it recorded the **user_id**, **rating**, and **timestamp** listed in that row. In this case, it found 452 matching rows.
- The second **movie_id** in movies is 2. Again, Pandas did a search of ratings and found 131 matching rows.
- This process was repeated for all of the remaining rows in movies.

At the end of the process, the movie_ratings DataFrame is created, which contains the two columns from movies (**movie_id** and **title**) and the three other colums from ratings (**user_id**, **rating**, and **timestamp**).

- **movie_id** 1 and its **title** are listed 452 times, next to the **user_id**, **rating**, and **timestamp** for each of the 452 matching ratings.
- **movie_id** 2 and its **title** are listed 131 times, next to the **user_id**, **rating**, and **timestamp** for each of the 131 matching ratings.
- And so on, for every movie in the dataset.

In [11]:
print( movies.shape)
print(ratings.shape)
print(movie_ratings.shape)

(1682, 2)
(100000, 4)
(100000, 5)


Notice the shapes of the three DataFrames:

- There are 1682 rows in the movies DataFrame.
- There are 100000 rows in the ratings DataFrame.
- The `merge` function resulted in a movie_ratings DataFrame with 100000 rows, because every row from ratings matched a row from movies.
- The movie_ratings DataFrame has 5 columns, namely the 2 columns from movies, plus the 4 columns from ratings, minus the 1 column in common.

By default, the `merge` function joins the DataFrames using all column names that are in common (**movie_id**, in this case). The [documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html) explains how you can override this behavior.

## Four Types of Joins

There are actually four types of joins supported by the Pandas `merge` function. Here's how they are described by the documentation:

- **inner:** use intersection of keys from both frames (SQL: inner join)
- **outer:** use union of keys from both frames (SQL: full outer join)
- **left:** use only keys from left frame (SQL: left outer join)
- **right:** use only keys from right frame (SQL: right outer join)

The default is the "inner join", which was used when creating the movie_ratings DataFrame.

It's easiest to understand the different types by looking at some simple examples:

### Example DataFrames A and B

In [12]:
A = pd.DataFrame({'color': ['green', 'yellow', 'red'], 'num':[1, 2, 3]})
A

color  num
0   green    1
1  yellow    2
2     red    3

In [13]:
B = pd.DataFrame({'color': ['green', 'yellow', 'pink'], 'size':['S', 'M', 'L']})
B

color size
0   green    S
1  yellow    M
2    pink    L

### Inner join

Only include observations found in both A and B:

In [14]:
pd.merge(A, B, how='inner')

color  num size
0   green    1    S
1  yellow    2    M

### Outer join

Include observations found in either A or B:

In [15]:
pd.merge(A, B, how='outer')

color  num size
0   green  1.0    S
1  yellow  2.0    M
2     red  3.0  NaN
3    pink  NaN    L

### Left join

Include all observations found in A:

In [16]:
pd.merge(A, B, how='left')

color  num size
0   green    1    S
1  yellow    2    M
2     red    3  NaN

### Right join

Include all observations found in B:

In [17]:
pd.merge(A, B, how='right')

color  num size
0   green  1.0    S
1  yellow  2.0    M
2    pink  NaN    L

# Extra Fun

## get top 5 movies with:
- best avg ratings
- most number of ratings
- best avg rating with most number of ratings


In [18]:
movie_ratings.groupby('title')['rating'].mean().sort_values(ascending=False)

title
Marlene Dietrich: Shadow and Light (1996)                 5.000000
Prefontaine (1997)                                        5.000000
Santa with Muscles (1996)                                 5.000000
Star Kid (1997)                                           5.000000
Someone Else's America (1995)                             5.000000
Entertaining Angels: The Dorothy Day Story (1996)         5.000000
Saint of Fort Washington, The (1993)                      5.000000
Great Day in Harlem, A (1994)                             5.000000
They Made Me a Criminal (1939)                            5.000000
Aiqing wansui (1994)                                      5.000000
Pather Panchali (1955)                                    4.625000
Anna (1996)                                               4.500000
Everest (1998)                                            4.500000
Maya Lin: A Strong Clear Vision (1994)                    4.500000
Some Mother's Son (1996)                                

In [19]:
movie_ratings.groupby('title')['rating'].count().sort_values(ascending=False)

title
Star Wars (1977)                                     583
Contact (1997)                                       509
Fargo (1996)                                         508
Return of the Jedi (1983)                            507
Liar Liar (1997)                                     485
English Patient, The (1996)                          481
Scream (1996)                                        478
Toy Story (1995)                                     452
Air Force One (1997)                                 431
Independence Day (ID4) (1996)                        429
Raiders of the Lost Ark (1981)                       420
Godfather, The (1972)                                413
Pulp Fiction (1994)                                  394
Twelve Monkeys (1995)                                392
Silence of the Lambs, The (1991)                     390
Jerry Maguire (1996)                                 384
Chasing Amy (1997)                                   379
Rock, The (1996)         

In [20]:
movie_ratings_avg = \
    movie_ratings.groupby('title')['rating'].mean().sort_values(ascending=False)
movie_ratings_avg = movie_ratings_avg.reset_index()
movie_ratings_avg.head()

title  rating
0  Marlene Dietrich: Shadow and Light (1996)      5.0
1                          Prefontaine (1997)     5.0
2                   Santa with Muscles (1996)     5.0
3                             Star Kid (1997)     5.0
4               Someone Else's America (1995)     5.0

In [21]:
movie_ratings_count = \
    movie_ratings.groupby('title')['rating'].count().sort_values(ascending=False)
movie_ratings_count = movie_ratings_count.reset_index()
movie_ratings_count.columns = ['title', 'count']
movie_ratings_count.head()

title  count
0           Star Wars (1977)    583
1             Contact (1997)    509
2               Fargo (1996)    508
3  Return of the Jedi (1983)    507
4           Liar Liar (1997)    485

In [22]:
pd.merge(movie_ratings_avg, movie_ratings_count).head()

title  rating  count
0  Marlene Dietrich: Shadow and Light (1996)      5.0      1
1                          Prefontaine (1997)     5.0      3
2                   Santa with Muscles (1996)     5.0      2
3                             Star Kid (1997)     5.0      3
4               Someone Else's America (1995)     5.0      1

In [26]:
pd.merge(movie_ratings_avg, movie_ratings_count).sort_values(by=['count', 'rating'], ascending=[False, False])

title    rating  count
23                                     Star Wars (1977)  4.358491    583
292                                      Contact (1997)  3.803536    509
64                                         Fargo (1996)  4.155512    508
129                           Return of the Jedi (1983)  4.007890    507
837                                    Liar Liar (1997)  3.156701    485
412                         English Patient, The (1996)  3.656965    481
597                                       Scream (1996)  3.441423    478
236                                    Toy Story (1995)  3.878319    452
428                                Air Force One (1997)  3.631090    431
598                       Independence Day (ID4) (1996)  3.438228    429
40                       Raiders of the Lost Ark (1981)  4.252381    420
34                                Godfather, The (1972)  4.283293    413
101                                 Pulp Fiction (1994)  4.060914    394
298                               Twelve Monkeys (1995)  3.798469    392
32                     Silence of the Lambs, The (1991)  4.289744    390
361                                Jerry Maguire (1996)  3.710938    384
260                                  Chasing Amy (1997)  3.839050    379
372                                    Rock, The (1996)  3.693122    378
49                      Empire Strikes Back, The (1980)  4.204360    367
409                     Star Trek: First Contact (1996)  3.660274    365
45                                       Titanic (1997)  4.245714    350
264                           Back to the Future (1985)  3.834286    350
700                          Mission: Impossible (1996)  3.313953    344
114                                Fugitive, The (1993)  4.044643    336
208           Indiana Jones and the Last Crusade (1989)  3.930514    331
427        Willy Wonka and the Chocolate Factory (1971)  3.631902    326
61                           Princess Bride, The (1987)  4.172840    324
256                                 Forrest Gump (1994)  3.853583    321
98               Monty Python and the Holy Grail (1974)  4.066456    316
871                                   Saint, The (1997)  3.123418    316
...                                                 ...       ...    ...
1630                            Crude Oasis, The (1995)  1.000000      1
1631                            Leopard Son, The (1996)  1.000000      1
1634                     Symphonie pastorale, La (1946)  1.000000      1
1635                                 Baton Rouge (1988)  1.000000      1
1637         Tigrero: A Film That Was Never Made (1994)  1.000000      1
1638                               Office Killer (1997)  1.000000      1
1639                                       T-Men (1947)  1.000000      1
1640                      Terror in a Texas Town (1958)  1.000000      1
1641         Nobody Loves Me (Keiner liebt mich) (1994)  1.000000      1
1642                               The Courtyard (1995)  1.000000      1
1643                            King of New York (1990)  1.000000      1
1644                                    Liebelei (1933)  1.000000      1
1645                        Stefano Quantestorie (1993)  1.000000      1
1646                          Hostile Intentions (1994)  1.000000      1
1647                              Pharaoh's Army (1995)  1.000000      1
1648                                New Age, The (1994)  1.000000      1
1649                     Hungarian Fairy Tale, A (1987)  1.000000      1
1650  Death in the Garden (Mort en ce jardin, La) (1...  1.000000      1
1651             Promise, The (Versprechen, Das) (1994)  1.000000      1
1652      I, Worst of All (Yo, la peor de todas) (1990)  1.000000      1
1653                                    Power 98 (1995)  1.000000      1
1654  Police Story 4: Project S (Chao ji ji hua) (1993)  1.000000      1
1655                               Low Life, The (1994)  1.000000      1
1657               Invitation, The 

# Can we achieve this in just 1 line?